### ISSUE TITLE
Restructure C-Sources directory so that it is according to Modelica specification

### ISSUE THREAD

**Issue by modelica-trac-importer** [modelica/Modelica#1021]
**Reported by otter on 13 Mar 2013 11:11 UTC**
Since some time the recommended structure to store C-sources and libraries is defined in the Modelica specification. Adapt Modelica\Resources\C-sources so that it is according to the specification (ModelicaExternalC library).

----

Migrated-From: https://trac.modelica.org/Modelica/ticket/1021

**Comment by modelica-trac-importer** [comment]
**Comment by otter on 26 Mar 2013 13:26 UTC**
In the Modelica specification only the include and (binary-) libraries are standardized, not the C-sources. Since currently no (binary) ModelicaExternalC library is shipped with MSL, the current structure for the C-sources is o.k. and therefore I did not change this.

How to continue is unclear to me and should be discussed in the next MAP-LIB web meeting. Either, just keep MSL as it is, or:

 1. Provide binary objects for the "Lapack" and "ModelicaExternalC" libraries for linux and windows in the Resources\Library directory.
 1. Standardize the handling of C-sources in Modelica 3.2 rev. 2, in order that a compiler can automatically translate the C-souces of the Lapack and ModelicaExternalC libraries to the desired target system (such as dSpace).
 1. Adapt the C-sources directory of MSL according to (3)

**Comment by modelica-trac-importer** [comment]
**Modified by otter on 26 Mar 2013 13:28 UTC**

**Comment by modelica-trac-importer** [comment]
**Comment by sjoelund&#46;se on 26 Mar 2013 18:44 UTC**
There are some options here. One is already in the specification. That uses the Include annotation also for C sources. The following finds the file `.../Modelica 3.2.1/Resources/Include/fn.c` (which I added to my local copy).
```mo
model M
  function fn
    output Integer i;
  external "C" annotation(IncludeDirectory="modelica://Modelica/Resources/Include" /* Added because this function is not part of Modelica */,Include="#include <fn.c>");
  end fn;
  Integer i = fn();
  annotation(uses(Modelica(version="3.2.1")));
end M;
```

If you use this approach you need to guard the functions just like you do with any includes (so you only add them once). It would probably be good to specify that it is only included once for simulation model (so if you link several C-files together, they do not include multiple versions of the C sources).

I dislike option (1) since you will not be able to ship versions that work well on all tools. Especially not shipping yet another lapack library. The one installed by the OS is probably better tuned than the version you would ship, especially if you use something like ATLAS.

I also slightly dislike the Include option I used above (which is sort of the same as (2)) since it means you need to recompile the library every time you use it. You then also need to consider the additional libraries that need to be linked on some platforms (libm on Linux, probably is already linked in, but might not be).

(2) could of course also be a standardized location of C sources so a tool could compile the library only once if needed (and then simply use Library="LibraryName" or something in an annotation as a predefined name for the C sources...).

**Comment by modelica-trac-importer** [comment]
**Comment by otter on 20 May 2013 07:47 UTC**
Fixed in 921074750b682aa80ce1432d7032d69372ad4a08:

As pointed out, in the Modelica specification the handling of C sources is not specified (only includes and object libraries). The issues currently are ModelicaExternalC and the LAPACK library. For both libraries it is not practical to include the C-code whenever a model is compiled. As discussed in a web/phone meeting, there seems to be agreement that this is a vendor issue: The source code for ModelicaExternalC is provided under Resources\C-Sources and the vendor is responsible to build a ModelicaExternalC library of it. The source code for LAPACK is not provided with MSL but the vendor is responsible to provide an appropriate library.

I added a _readme.txt file under Modelica/Resources/C-Sources that explains this approach.

**Comment by modelica-trac-importer** [comment]
**Modified by otter on 20 May 2013 07:53 UTC**
